caas/kubernetes/provider/k8s.go

```diff
@@ -155,9 +178,132 @@ func (k *kubernetesClient) EnsureOperator(appName, agentPath string, config *caa
 	return nil
 }
 
+// maybeGetOperatorStorageClass looks for a storage class to use when creating
+// a persistent volume for an operator.
+func (k *kubernetesClient) maybeGetOperatorStorageClass() (*storage.StorageClass, error) {
+	// First try looking for a storage class with a Juju label.
+	storageClasses, err := k.StorageV1().StorageClasses().List(v1.ListOptions{
+		LabelSelector: operatorStorageSelector(),
+	})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	logger.Debugf("available storage classes: %#v", storageClasses.Items)
+	if len(storageClasses.Items) > 0 {
+		return &storageClasses.Items[0], nil
+	}
+
+	// Second look for the default storage class, if defined.
+	storageClasses, err = k.StorageV1().StorageClasses().List(v1.ListOptions{})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	for _, sc := range storageClasses.Items {
+		if v, ok := sc.Annotations["storageclass.kubernetes.io/is-default-class"]; ok && v != "false" {
+			logger.Debugf("using default storage class: %v", sc.Name)
+			return &sc, nil
+		}
+	}
+	return nil, errors.NotFoundf("storage class for operator storage")
+}
+
+func operatorVolumeClaim(appName string) string {
+	return fmt.Sprintf("%v-operator-volume-claim", appName)
+}
+
+// maybeGetOperatorVolume attempts to create a persistent volume to use with an operator.
+func (k *kubernetesClient) maybeGetOperatorVolume(appName, operatorVolumeSize string) (*core.Volume, error) {
+	// We create a volume using a persistent volume claim.
+	// First, attempt to get any previously created claim for this app.
+	pvClaims := k.CoreV1().PersistentVolumeClaims(k.namespace)
+	pvcName := operatorVolumeClaim(appName)
+	_, err := pvClaims.Get(pvcName, v1.GetOptions{})
+	if err != nil && !k8serrors.IsNotFound(err) {
+		return nil, errors.Trace(err)
+	}
+	if err != nil {
```

Comment:
This would be easier to follow with an early return of the volume with the existing pvcName (maybe using a func to construct the core.Volume to avoid duplication).

Reply:
done